### PR TITLE
feat: Add snap_start functionality

### PR DESCRIPTION
**Pull request by darrenswhite**
## How Has This Been Tested?
- [ ] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [ ] I have tested and validated these changes using one or more of the provided `examples/*` projects

- [ ] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

File: versions.tf [hunk 1]
```diff
@@ -4,7 +4,7 @@ terraform {
   required_providers {
```
Comment (mmichna) [inline versions.tf:4]: maybe is better upgrade also terraform version: `required_version = ">= 1.3"`

File: variables.tf [hunk 2]
```diff
@@ -217,6 +217,12 @@ variable "image_config_working_directory" {
   default     = null
 }
 
+variable "snap_start_apply_on" {
```
Comment (mmichna) [inline variables.tf:220]: maybe is better change it to:
```
variable "snap_start" {
  description = "(Optional) Snap start settings for low-latency startups"
  type        = bool
  default     = false
}
```

File: main.tf [hunk 3]
```diff
@@ -102,6 +102,14 @@ resource "aws_lambda_function" "this" {
     }
   }
 
+  dynamic "snap_start" {
```
Comment (mmichna) [inline main.tf:105]: there missing publish field. Maybe is better use this:
```
  dynamic snap_start {
    for_each = var.snap_start ? [true] : []
    content {
      apply_on = "PublishedVersions"
    }
  }
  publish = var.snap_start
```

**Comment by faxx1080** [comment]
Hi! I wanted to confirm, does this include publishing versions? I have heard from AWS that versions are required for full use of Snap Start.

See: https://docs.aws.amazon.com/lambda/latest/dg/snapstart.html

> You can use SnapStart only on [published function versions](https://docs.aws.amazon.com/lambda/latest/dg/configuration-versions.html#configuration-versions-config) and [aliases](https://docs.aws.amazon.com/lambda/latest/dg/configuration-aliases.html) that point to versions. You can't use SnapStart on a function's unpublished version ($LATEST).

**Comment by darrenswhite** [comment]
> Hi! I wanted to confirm, does this include publishing versions? I have heard from AWS that versions are required for full use of Snap Start.
> 
> See: https://docs.aws.amazon.com/lambda/latest/dg/snapstart.html
> 
> > You can use SnapStart only on [published function versions](https://docs.aws.amazon.com/lambda/latest/dg/configuration-versions.html#configuration-versions-config) and [aliases](https://docs.aws.amazon.com/lambda/latest/dg/configuration-aliases.html) that point to versions. You can't use SnapStart on a function's unpublished version ($LATEST).

this PR does not include publishing versions. you can use the already existing variable `publish` to enable publishing a new version here https://github.com/terraform-aws-modules/terraform-aws-lambda/blob/master/variables.tf#L125

**Comment by darrenswhite** [comment]
Replaced with new PR #406.

**Comment by darrenswhite** [comment]
@mmichna I have updated the new PR with your comments.

@faxx1080 the new PR will automatically publish a new version when snap start is enabled.

**Comment by faxx1080** [comment]
Hi @darrenswhite 

Thank you! However I am not sure what this does, we can already use "publish = true" for snapstart.

My thought is what happens to older versions of the function?

Currently I wrote a Perl script that uses the AWS CLI to enumerate versions not tagged and deletes those one by one. I use a null terraform provider to invoke the Perl script.

**Pull request closed without merging**